core/vm/lfvm/instructions.go

```diff
@@ -511,16 +620,35 @@ func opGasPrice(c *context) {
 func opBalance(c *context) {
 	slot := c.stack.peek()
 	address := common.Address(slot.Bytes20())
+	err := gasEip2929AccountCheck(c, address)
+	if err != nil {
+		c.status = OUT_OF_GAS
+		return
+	}
 	slot.SetFromBig(c.evm.StateDB.GetBalance(address))
 }
 
 func opSelfbalance(c *context) {
 	c.stack.pushEmpty().SetFromBig(c.evm.StateDB.GetBalance(c.contract.Address()))
 }
 
+func opBaseFee(c *context) {
+	baseFee, _ := uint256.FromBig(c.evm.Context.BaseFee)
+	c.stack.push(baseFee)
+}
+
 func opSelfdestruct(c *context) {
+	if c.stack.stack_ptr < 1 {
+		c.status = ERROR
+		return
+	}
+	gasfunc := gasSelfdestruct
+	if c.evm.ChainConfig().IsBerlin(c.evm.Context.BlockNumber) {
+		gasfunc = gasSelfdestructEIP2929
+	}
 	// even death is not for free
-	if !c.UseGas(gasSelfdestruct(c)) {
+	if !c.UseGas(gasfunc(c)) {
+		c.status = OUT_OF_GAS
```

Comment:
nit: UseGas is already updating the c.status to OUT_OF_GAS.

Reply:
done